README.md
# Firmware for a DIY Reflow Hot-Plate like "[Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/)"
![One of my "Tim's Hot-Plate" builds](assets/images/TimsHotPlate-1.jpg)

A feature-rich reflow Hot-Plate firmware, i.e. usable for "[Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/)"

## Introduction

I was in the need for a reflow soldering Hot-Plate.

After evaluating a couple of good projects, I decided to build "[Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/)" as it had a very detailed and quite good description, as well as I like the way how he did it with less or minimal external parts.

It was quickly build up and it worked quite good so that I could do my reflow soldering project for which I needed the Hot-Plate.

But: I decided for a larger (and more powerfull) 400W PTC as well as I mounted the Thermocouple in a different way as Tim did/described in his description.
The result was, that I had to struggle with quite large overshooting (up to +40°C).
Not a big deal as it was fairly easiable to compensate this overshooting by choosing a smaller target temp before reaching the real target temp. 

However, during the reflow of a coulpe of prototype PCB's, the overshooting workaround started to annoy me, as well as I got a couple of ideas (while rotating the knob and watching/waiting for the target temperature).

As we all have some kind of fetish, and mine seem to be that I tend to over-optimize things others are fine with, I decided trying to build a more feature-rich firmware alternative.

## Features

### Setup

Built-in "Setup" within Hot-Plate user interface

![Built-in Setup](assets/images/Setup-1.jpg)

### Reflow Profile

"Manual" or (open-end) "Reflow-Profile" Mode, with built-in reflow profiles for low-temp. solder paste Sn42/Bi57.6/Ag0.4, as well as high-temp. Sn96.5/Ag3.0/Cu0. After (as well as during) the reflow profile time targets, the user may adapt the target temp. (every 10 seconds). At profile end, the last target temp remain active and the user has to stop the profile manually when everything reflowed correctly.

![Reflow Profiles](assets/images/ReflowProfiles.jpg)
![Reflow Profile Start](assets/images/ReflowProfile-Start.jpg)
![Running Reflow Profile](assets/images/ReflowProfile-1.jpg)

### [PID Tuner](https://pidtuner.com) support

Simply start an automatic [step response](https://en.wikipedia.org/wiki/Step_response) run, with a connected serial console.
Afterwards copy the resulting data into [PID Tuner](https://pidtuner.com) and (auto-)tune your PID constants.
BTW: This step response can also be used to determine the overshot of the tuned target temperatures which might be used (at least as indication) for the [BangON](#bangonbangoff) setting. 

![Setup BangON](assets/images/PIDTuner-Start.jpg)
![Setup BangOFF](assets/images/PIDTuner-Heat.jpg)
![Setup BangOFF](assets/images/PIDTuner-Settle.jpg)
![Setup BangOFF](assets/images/PIDTuner-Output-1.jpg)

### PID constants

Either you know your PID constants, evaluate them the hard way "by hand" or use [PID Tuner support](#pid-tuner-support) for assistance in PID constant determination. 

![Setup PID constant Kp](assets/images/Setup-PID-Kp.jpg)
![Setup PID constant Kd](assets/images/Setup-PID-Kd.jpg)
![Setup PID constant Ki](assets/images/Setup-PID-Ki.jpg)

### BangON/BangOff

Quick start/Overshoot reduction settings and functionality. Mainly for bad/non-configuraed PID controller constants.
A good configured PID controller normally does not need BangON/BangOFF, but PID controller constant determination is complicated and time consuming. See [PID Tuner support](#pid-tuner-support) for assistance in PID constant determination.

![Setup BangON](assets/images/Setup-BangON.jpg)
![Setup BangOFF](assets/images/Setup-BangOFF.jpg)

### SSR type

Active-low/high configurable within buit-in setup

![Setup SSR Type](assets/images/Setup-SSR.jpg)

### Save

Save of settings to EEPROM possible within built-in setup 

![Setup Save](assets/images/Setup-Save.jpg)

## Requirements

Assembled DIY Hot Plate like described in "[Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/)" which is based on the following components

- ATmega328 like Arduino Nano 3.x
- Some kind of cheap AC-Mains powered PTC heater (200-400W)
- SSR (active- high or low) which has enough power to switch your selected PTC
- Convenient K-Type Thermocouple
- MAX6675 Module
- SSD1306 0.96" I2C OLED 128x64 Display
- Rotary EC11 Encoder

## Roadmap

- [X] Step response output for easier PID constant determination by the help of [PID Tuner](https://pidtuner.com)
- [ ] Some kind of PID loop tuning/calibration
- [X] Ramp-up determination to identify ~~PTC time and TC delay~~ (BangON calibration)
- [ ] Progressive rotary switching for quicker +/- move
- [ ] C/F unit selection (if someone is interested in)
- [ ] Larger display so that I don't need my glasses (old man wish)

## Known Bugs/Issues

- [X] I've the impression that the Ki constant of the (currently) used [AutoPID](https://r-downing.github.io/AutoPID/) doesn't behave as expected -> Now fixed

## Technology

The code structure is [PlatformIO/VSCode](https://platformio.org/) based (no Ardino-IDE test or support).

## Installing

1. Upload .hex or .elf file WITHOUT HAVING AC-MAINS CONNECTED
1. Long-press (> 0.5s) to get into built-in setup
1. Choose "SSR Type"
1. Select your SSR type (Active-Low/High). In the case  of [Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/) it will be "Active-Low"
1. Choose "Save & Quit"
1. Plug-in AC-Mains. Happy reflow soldering...

## Usage

- Rotate knob for +/- or up/down
- Push knob to stop (or start profile if it's displayed on screen)
- Long-Push (> 0.5s) to get into built-in setup

## Built With

* [Tim's Hot Plate](https://www.instructables.com/Tims-Hot-Plate/) - Quite cool DIY reflow Hot Plate Project @ Instructables
* [U8g2](https://github.com/olikraus/u8g2) - Excellent and powerful Library for monochrome displays, version 2
* [MAX6675](https://github.com/adafruit/MAX6675-library) - Arduino library for interfacing with MAX6675 thermocouple amplifier
* [CRC32](https://github.com/bakercp/CRC32) - An Arduino library for calculating a CRC32 checksum
* [AutoPID](https://github.com/r-downing/AutoPID) - Fairly good documented (as well as feature rich and small) Arduino AutoPID library 
* [PID Tuner](https://pidtuner.com) - Will help you to finde reasonable PID constants

## Contributing

Please read [CONTRIBUTING.md](./CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning.

For the versions available, please see [changelog](CHANGELOG.md) file or the [tags on this repository](https://github.com/Apehaenger/Another-HotPlate-Firmware/tags). 

## Authors

* Jörg Ebeling - *Initial work* - [Apehaenger](https://github.com/Apehaenger)

<!--
See also the list of [contributors](https://github.com/your/project/contributors) who participated in this project.
-->

## License

This project is licensed under the GNU General Public License 3 (GPL-3.0) - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [Unreleased]

### Added 

### Removed

- tbd
- something from [@someone](https://github.com/someone).

-->

## [0.5.0] - 2022-11-27

### Added 

- Added [PID Tuner](https://pidtuner.com) output
- Added average temperatures for more stable temperature display/handling 
- Added "Max. Temp" setting (which is required for PID Tuner steps)
- Added Changelog

### Changed

- Enlarged PID constant length from 3 to 4 int digits
- Improved some internal var names
- Optimized Thermocouple class and implemented average temperature
- Moved profile handling into a separate class
- Resturctured handling of objects via globals (instead of space wasting pointer/reference parameter)
- Changed initial PID-Tuner wait time and saved some resources

### Fixed

- Fixed possibliy wrong PID derivate calculation as discussed [here](https://github.com/r-downing/AutoPID/issues/4)


## [0.4.0] - 2022-11-08

### Added

- "Manual" or (open-end) "Reflow-Profile" Mode
- Reflow profiles for (low temp.) "Sn42/Bi57.6/Ag0.4" as well as (high temp.) "Sn96.5/Ag3.0/Cu0"

### Changed

- Moved most code into separate classes for higher flexibility and safer var/pointers
- Stripped down fonts, to get some more flash space

## 0.3.0 - 2022-10-26

### Added

- Built-in Setup
- Adjustabled PID constants
- BangON and BangOFF values
- Config storage into EEPROM

[unreleased]: https://github.com/Apehaenger/Another-HotPlate-Firmware/compare/v0.5.0...develop
[0.5.0]: https://github.com/Apehaenger/Another-HotPlate-Firmware/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/Apehaenger/Another-HotPlate-Firmware/releases/v0.4.0


src/Display.cpp
/*
 * This file is part of the Another-Reflow-HotPlate-Firmware project (https://github.com/Apehaenger/Another-Reflow-HotPlate-Firmware).
 * Copyright (c) 2022 Jörg Ebeling
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include "main.hpp"
#include "Display.hpp"
#include "config.hpp"
#include "../assets/fonts/my_u8g2_font_7x13B.hpp"
#include "../assets/fonts/my_u8g2_font_open_iconic_embedded_2x.hpp"
#include "../assets/fonts/my_u8g2_font_fur20.hpp"

/*
 * 0.96" Display. Two colored version has:
 *   16 pixel i.e. yellow
 *    1 non-counted empty pixel row
 *   48 pixel i.e. blue
 */

Display::Display(uint16_t interval_ms, Thermocouple *TcPtr, Hotplate *HotPtr) : Runnable(interval_ms),
                                                                                _TcPtr(TcPtr),
                                                                                _HotPtr(HotPtr),
                                                                                u8g2(U8G2_R0, /* reset=*/U8X8_PIN_NONE)
{
}

void Display::setup()
{
    u8g2.begin(ROTARY_S_PIN, ROTARY_A_PIN, ROTARY_B_PIN);
    u8g2.clear();
}

void Display::mainScreen()
{
    // Check if we need to make the expensive display redraw
    if (_TcPtr->getTemperatureAverage() == _lastTemp &&
        _HotPtr->getSetpoint() == _lastTarget &&
        _HotPtr->getPower() == _lastPower &&
        _HotPtr->getProfileTimePosition() == _lastProfileTimePosition)
    {
        return;
    }

    char cbuf[12]; // Longest entry length = "Target: 123\0"
    String s = "";
    u8g2_uint_t y;

    u8g2.firstPage();
    do
    {
        // Standard (small font)
        setStdFont();

        // Profile / Mode
        u8g2.drawStr(0, 13, _HotPtr->profile2str[Config::active.profile]);

        // 2nd row
        y = 27;
        if (Config::active.profile != Hotplate::Profile::Manual &&
            _HotPtr->getControllerState() == Hotplate::ControllerState::off)
        {
            s = "Push to start";
            u8g2.drawStr((u8g2.getDisplayWidth() - u8g2.getStrWidth(s.c_str())) / 2, y, s.c_str());
        }
        else
        {
            // Target temperature
            _lastTarget = _HotPtr->getSetpoint();
            sprintf(cbuf, "Target: %3d", _lastTarget);
            u8g2.drawFrame(u8g2.getStrWidth(cbuf) - (3 * 7) - 2, 16, (3 * 7) + 4, 13);
            u8g2.drawStr(0, y, cbuf);
            if (Config::active.profile != Hotplate::Profile::Manual)
            {
                sprintf(cbuf, "%3ds", _HotPtr->getProfileTimePosition());
                u8g2.drawStr(85, y, cbuf);
            }
        }

        // Controller state
        switch (_HotPtr->getControllerState())
        {
        case Hotplate::ControllerState::bangOn:
            s = "BangON";
            u8g2.setFontMode(0);
            u8g2.setDrawColor(1);
            u8g2.drawBox(0, 29, u8g2.getStrWidth(s.c_str()) + 2, 13);
            u8g2.setDrawColor(0);
            u8g2.drawStr(1, 40, s.c_str());
            break;
        case Hotplate::ControllerState::pid:
            s = "PID ";
            s += _HotPtr->getOutput();
            s += "/";
            s += Config::active.pid_pwm_window_ms;
            u8g2.drawStr(1, 40, s.c_str());
            break;
        case Hotplate::ControllerState::bangOff:
            u8g2.drawStr(1, 40, "BangOFF");
            break;
        default:
            break;
        }

        // Unit
        u8g2.setDrawColor(1);
        u8g2.drawUTF8(108, 62, "°C");

        // ----- Large font stuff -----
        u8g2.setFont(my_u8g2_font_fur20);

        // Temperature
        dtostrf(_TcPtr->getTemperatureAverage(), 5, 1, cbuf);
        s = String(cbuf);
        s.concat(" °C");
        u8g2.drawUTF8(35, 64, s.c_str());
        _lastTemp = _TcPtr->getTemperatureAverage();

        // SSR Power
        _lastPower = _HotPtr->getPower();
        if (_lastPower)
        {
            u8g2.setFont(my_u8g2_font_open_iconic_embedded_2x);
            u8g2.drawStr(5, 62, "C"); // Power symbol = 67 = C
        }

    } while (u8g2.nextPage());
}

void Display::changeUiMode(uiMode nextUi)
{
    uiM = nextUi;
}

void Display::userInterfaceInputDouble(const char *title, const char *pre, double *value, uint8_t numInt, uint8_t numDec, const char *post)
{
    uint8_t iV;
    String valueStr = "", prePart, postPart;

    // Double -> String
    char cbuf[numInt + numDec + 1];
    if (*value == 0)
    {
        for (uint8_t i = 0; i < numInt; i++)
            valueStr += "0";
        valueStr += ".";
        for (uint8_t i = 0; i < numDec; i++)
            valueStr += "0";
    }
    else
    {
        dtostrf(*value, numInt + numDec + 1, numDec, cbuf);
        valueStr = String(cbuf);
    }

    // Serial.print("valueStr: ");
    // Serial.println(valueStr);

    // Loop over als string chars
    for (uint8_t i = 0; i < valueStr.length(); i++)
    {
        if (valueStr[i] == '.')
            continue;

        if (valueStr[i] == ' ')
            valueStr[i] = '0';

        prePart = pre;
        prePart += valueStr.substring(0, i);
        prePart += "[";

        iV = valueStr[i] & 0x0f;

        postPart = "]";
        postPart += valueStr.substring(i + 1);

        /*
        Serial.print("i: ");
        Serial.print(i, HEX);
        Serial.print(", pre: '");
        Serial.print(pre);
        Serial.print("', prePart: '");
        Serial.print(prePart);
        Serial.print("', edit digit ");
        Serial.print(iV, HEX);
        Serial.print(", postPart: '");
        Serial.print(postPart);
        Serial.println("'");
        */

        if (u8g2.userInterfaceInputValue(title, prePart.c_str(), &iV, 0, 9, 1, postPart.c_str()) == 0)
            return; // We don't have a "home" (escape) button yet
        valueStr[i] = iV + 0x30;
    }
    *value = valueStr.toDouble();
}

void Display::inputBangValues()
{
    u8g2.firstPage();
    do
    {
        setStdFont();
        // FIXME: The following should go into a loop with a mapped "Kx" string and config var
        if (u8g2.userInterfaceInputValue("Bang-ON until\ntarget-temp", "minus ", &Config::active.pid_bangOn_temp_c, 0, 255, 3, " °C") == 0)
            return; // We don't have a "home" (escape) button yet
        if (u8g2.userInterfaceInputValue("Bang-OFF at\ntarget-temp", "plus ", &Config::active.pid_bangOff_temp_c, 0, 255, 3, " °C") == 0)
            return; // We don't have a "home" (escape) button yet
    } while (u8g2.nextPage());
}

void Display::inputPidConstants()
{
    u8g2.firstPage();
    do
    {
        setStdFont();
        // FIXME: The following should go into a loop with a mapped "Kx" string and config var
        const char *title = "Select\nPID constant";
        userInterfaceInputDouble(title, "Kp = ", &Config::active.pid_Kp, 3, 1, "");
        userInterfaceInputDouble(title, "Ki = ", &Config::active.pid_Ki, 3, 1, "");
        userInterfaceInputDouble(title, "Kd = ", &Config::active.pid_Kd, 3, 1, "");
    } while (u8g2.nextPage());
}

void Display::setStdFont()
{
    // u8g2.setFont(my_u8g2_font_8x13B);
    u8g2.setFont(my_u8g2_font_7x13B);
}

void Display::inputSsrType()
{
    u8g2.firstPage();
    do
    {
        setStdFont();
        uint8_t sel = u8g2.userInterfaceSelectionList("Setup SSR Type", (Config::active.ssr_active_low ? 1 : 2), "Active Low\nActive High");
        switch (sel)
        {
        case 1: // Active Low
            Config::active.ssr_active_low = true;
            break;
        case 2: // Active High
            Config::active.ssr_active_low = false;
            break;
        }
    } while (u8g2.nextPage());
}

void Display::inputReflowProfile()
{
    u8g2.firstPage();
    do
    {
        setStdFont();
        String modeList = "";

        for (uint8_t i = 0; i < sizeof(Hotplate::profile2str) / sizeof(char *); ++i)
        {
            if (i > 0)
                modeList += "\n";
            modeList += _HotPtr->profile2str[i];
        }

        uint8_t sel = u8g2.userInterfaceSelectionList("Setup Profile", Config::active.profile + 1, modeList.c_str());
        Config::active.profile = static_cast<Hotplate::Profile>(sel - 1);
    } while (u8g2.nextPage());
}

void Display::setupScreen()
{
    u8g2.firstPage();
    do
    {
        setStdFont();
        String s = "Setup (";
        s += VERSION_TEXT;
        s += ")";

        /* Setup
         * -----
         * Reflow Profile
         * Display unit
         * SSR Type
         * PID Coefficients
         * Calibration
         * Save to EEPROM
         * Quit
         */
        uint8_t sel = u8g2.userInterfaceSelectionList(s.c_str(), 1,
                                                      "Reflow Profile\n(Display unit)\nSSR Type\nPID constants\nBangBang\n(Calibration)\nLoad saved\nSave & Quit\nQuit");
        //                                                    1               2              3         4             5           6             7            8        9
        switch (sel)
        {
        case 1: // Reflow Profile
            inputReflowProfile();
            break;
        case 3: // SSR Type
            inputSsrType();
            break;
        case 4: // PID constants
            inputPidConstants();
            _HotPtr->updatePidGains();
            break;
        case 5: // BangBang values
            inputBangValues();
            break;
        case 7: // Load saved
            Config::load();
            break;
        case 8: // Save & Quit
            Config::save();
            changeUiMode(uiMode::Main);
            break;
        default:
            changeUiMode(uiMode::Main);
        }
    } while (u8g2.nextPage());
}

void Display::loop()
{
    switch (uiM)
    {
    case uiMode::Setup:
        setupScreen();
        /*
                        // Double -> String
                        char cbuf[10];
                        dtostrf(conf->pid_Kp, 8, 2, cbuf);
                        Serial.print("Before: ");
                        Serial.println(cbuf);

                        setStdFont();
                        inputPidConstants();

                        // Double -> String
                        dtostrf(conf->pid_Kp, 8, 2, cbuf);
                        Serial.print("After: ");
                        Serial.println(cbuf);
        */
        break;
    default:
        mainScreen();
    }
}